src/charm.py

```diff
@@ -113,85 +106,125 @@ def _on_jenkins_pebble_ready(self, event: PebbleReadyEvent) -> None:
+            self.unit.status = ops.BlockedStatus("Timed out waiting for Jenkins.")
             return
         except jenkins.JenkinsBootstrapError as exc:
             logger.error("Error installing plugins, %s", exc)
-            self.unit.status = BlockedStatus("Error installling plugins.")
+            self.unit.status = ops.BlockedStatus("Error installling plugins.")
             return
 
         try:
             version = jenkins.get_version()
         except jenkins.JenkinsError as exc:
             logger.error("Failed to get Jenkins version, %s", exc)
-            self.unit.status = BlockedStatus("Failed to get Jenkins version.")
+            self.unit.status = ops.BlockedStatus("Failed to get Jenkins version.")
             return
 
         self.unit.set_workload_version(version)
-        self.unit.status = ActiveStatus()
+        self.unit.status = ops.ActiveStatus()
 
-    def _on_get_admin_password(self, event: ActionEvent) -> None:
+    def _on_get_admin_password(self, event: ops.ActionEvent) -> None:
         """Handle get-admin-password event.
 
         Args:
             event: The event fired from get-admin-password action.
         """
-        if not self._jenkins_container.can_connect():
+        container = self.unit.get_container(self.state.jenkins_service_name)
+        if not container.can_connect():
             event.defer()
             return
-        credentials = jenkins.get_admin_credentials(self._jenkins_container)
+        credentials = jenkins.get_admin_credentials(container)
         event.set_results({"password": credentials.password})
 
-    def _on_update_status(self, _: UpdateStatusEvent) -> None:
-        """Handle update status event.
+    def _remove_unlisted_plugins(self, container: ops.Container) -> ops.StatusBase:
+        """Remove plugins that are installed but not allowed.
 
-        On Update status:
-        1. Update Jenkins patch LTS version if available.
-        2. Update apt packages if available.
+        Args:
+            container: The jenkins workload container.
+
+        Returns:
+            The unit status of the charm after the operation.
         """
-        if self.state.update_time_range and not self.state.update_time_range.check_now():
-            self.unit.status = ActiveStatus()
-            return
+        original_status = self.unit.status.name
+        try:
+            jenkins.remove_unlisted_plugins(plugins=self.state.plugins, container=container)
+        except (jenkins.JenkinsPluginError, jenkins.JenkinsError) as exc:
+            logger.error("Failed to remove unlisted plugin, %s", exc)
+            return ops.StatusBase.from_name(original_status, "Failed to remove unlisted plugin.")
+        except TimeoutError as exc:
+            logger.error("Failed to restart jenkins after removing plugin, %s", exc)
+            return ops.BlockedStatus("Failed to restart Jenkins (remove plugin)")
```

Comment:
The `(remove plugin)` could be confusing, is the intent to indicate why the restart was triggered?

Reply:
Yes, the two jenkins "management" functions `_remove_unlisted_plugins` and `_update_jenkins_version` both may block on failure to restart, so I thought such status message could help differentiate the two.
Perhaps "Failed to restart Jenkins after removing plugins" might be better?